### DISCUSSION TITLE
Type inference error on Connect::io

### DISCUSSION THREAD

**Discussion by tik** [ENQT-GmbH/remoc#22] [Q&A] [answered]
Hi,

this is not strictly a remoc problem but seems close enough, so I hope I can get some help over here.

I'm currently using remoc in a project and I really like it. Using the ConnectExt::provide() method to send a single trait object to a remote endpoint works great. What I want to do now is, to keep the base channel around and dynamically send one ore more other trait objects to the endpoint. Something like this:

```
let (conn, tx, rx) = remoc::Connect::io(remoc::Cfg::default(), io::stdin(), io::stdout()).await.unwrap();
tokio::spawn(conn);

//directly, when establishing the comnnection
let (trait_a_server, trait_a_client) = TraitAServerRefMut::<_, codec::Default>::new(self, 1);
let _ = tx.send(trait_a_client).await;
trait_a_server.serve().await;

...
// sometime later, when the remote endpoint requests it via TraitA
let (trait_b_server, trait_b_client) = TraitBServerRefMut::<_, codec::Default>::new(self, 1);
let _ = tx.send(trait_b_client).await;
trait_b_server.serve().await;
```

The problem is, that rustc can't infer the types of tx and rx (or inferes it to TraitA, when that object is send directly). What is
the idiomatic way to have more than one remote trait with remoc?

Thanks,
Tim

**Comment by surban** [comment, marked as answer]
Hi,

the `connect` method returns a base channel, which, like all channels, transfers objects of a fixed data type. The compiler correctly infers that data type to be of `TraitAClient` since this is the first thing you send over it.

The solution is to transfer other data types by other means, since remoc does not limit you to one channel. These other means can be further Remoc channels or parameters or return value of remote traits. 

In your comment you write that `remote endpoint requests it [TraitB] via TraitA`. You should be able to simply return `TraitBClient` from the trait method that is used to request it. There is no need to send it over the initial channel. 

If, for some reason, you need to exchange multiple data types over a single channel, the solution is to define a Rust enum that holds all the data types you want to transfer. Remoc channels and remote trait clients can be transferred as fields of user-defined enums and structs.

It would help if you would share a repository with your code. This way I could check if your approach can be simplified.

Hope that helps!
